Fix #465: CA1812: AvoidUninstantiatedInternalClasses

lgolding opened the pull request:
@dotnet/roslyn-analysis

lgolding commented on src/Analyzer.Utilities/Extensions/ISymbolExtensions.cs:
```diff
@@ -147,6 +147,30 @@ public static bool MatchFieldByName(this ISymbol member, INamedTypeSymbol type,
         {
             return member != null && member.Kind == SymbolKind.Field && member.MatchMemberByName(type, name);
         }
+
```
Move this method from the CA1716 analyzer into Analyzer.Utilities. All FxCop rules format the names of nested classes this way. Odds are there are other analyzers, already implemented, whose output wouldn't be exactly the same as FxCops when the fired against a nested class.

lgolding commented on src/Microsoft.Maintainability.Analyzers/Core/MicrosoftMaintainabilityAnalyzersResources.resx:
```diff
@@ -173,7 +172,7 @@
     <value>An instance of an assembly-level type is not created by code in the assembly.</value>
   </data>
   <data name="AvoidUninstantiatedInternalClassesMessage" xml:space="preserve">
-    <value>Avoid uninstantiated internal classes</value>
```
For some reason this message didn't have the correct text.

mavasani commented:
Wouldn't this rule be too noisy in presence of InternalsVisibleTo? I think FXCop handles this case: http://index/#FxCopSdk/InternalUtilities.cs,436

lgolding commented:
You're right. This was a case where I didn't think I needed to look at the FxCop details because I thought it was "obvious", and I was wrong. I'll add code to cover the other cases.

mavasani commented:
Also, you probably want to ignore types with the MEF export attributes - they wouldn't have an explicit instantiation. And Roslyn code is full of such types..

lgolding commented:
@mvasani You're right. Now, our POR is that in the initial implementation, we'll be as compatible with FxCop as possible, to avoid customers being surprised by new messages when they migrate from FxCop to the Roslyn-based equivalents (and also to ease testing: we want to be able to directly compare FxCop's output to the analyzers's output). In the next version, we'll consider what improvements we want to make.

We're tracking those changes in a document [Proposed FxCop changes in Roslyn](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/FxCopPort/Proposed%20FxCop%20rule%20changes%20in%20Roslyn.md). I'll add your suggestion to that document.

srivatsn commented:
In this case though, wouldn't filtering out MEF exporting types just reduce the number of new messages?

heejaechang commented:
+! if new implementation can reduce false positive we should do that. I think we should only worry about implementation being different when we shows more diagnostics regardless whether it is false positive or correct.

heejaechang commented:
is this PR broken? why are they so many changes and commits which seems not related to this PR?

lgolding commented:
Hi @heejaechang, I was doing this work on a branch (CA1812) that had become badly out of date with master. I merged the latest master into my branch and pushed the result. This PR is now repaired (but still not complete).

lgolding commented:
@mavasani @srivatsn @heejaechang Now handles both MEF 1 and MEF 2 exported classes (does not emit diagnostics for unused internal class if it is marked with either one of the two `Export` attributes).

@mavasani: Now looking at InternalsVisibleTo.

lgolding commented:
@srivatsn @mavasani @heejaechang 

OK, I made the change so that we don't emit any diagnostics if the assembly being compiled has any `InternalsVisibleTo` attributes. I also adjusted the unit tests to create the MEF-related and Configuration-related classes they need, so I removed the extra references I added to the test utilities project. So Manish's test is now re-enabled.

The new product code is the method `AssemblyExposesInternals` in the analyzer class.

mavasani commented on src/System.Collections.Immutable.Analyzers/UnitTests/System.Collections.Immutable.Analyzers.UnitTests.csproj:
```diff
@@ -68,6 +68,9 @@
     <Compile Include="DoNotCallToImmutableArrayOnAnImmutableArrayValueTests.Fixer.cs" />
   </ItemGroup>
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
```
Changes to project files seem unintentional - can you please revert them?

lgolding replied:
Fixed, pushed change.

mavasani commented on src/Analyzer.Utilities/Extensions/INamedTypeSymbolExtensions.cs:
```diff
@@ -124,7 +124,113 @@ public static bool HasFinalizer(this INamedTypeSymbol symbol)
+                .Any(m => m.IsFinalizer());
+        }
+
+        public static bool HasAttribute(this INamedTypeSymbol symbol, INamedTypeSymbol attribute)
```
Do we want to also include attributes on the base type/interfaces? Probably add a flag to control that? Regardless, can we add a comment so the consumer of the API is aware of the functionality?

lgolding replied:
I added an XML doc comment on the method and documented the existing behavior, which is that it does not find attributes on base types.

mavasani commented on src/Analyzer.Utilities/Extensions/ISymbolExtensions.cs:
```diff
@@ -151,6 +161,30 @@ public static bool MatchFieldByName(this ISymbol member, INamedTypeSymbol type,
+        /// <returns>
+        /// A string representing the name of the member in a format consistent with FxCop.
+        /// </returns>
+        public static string FormatMemberName(this ISymbol member)
```
I thought we decided to display only the simple member name in Roslyn diagnostic messages. Are we using this API in any place apart from diagnostic messages?

lgolding replied:
There are only four calls to this API, all in the context of `ReportDiagnostic`. But in any case, this method does not emit the namespace, it only emits the type name. I have tests for this, for example:

```
        // The following tests are just to ensure that the messages are formatted properly
        // for types within namespaces.
        [Fact]
        public void CA1812_CSharp_Diagnostic_UninstantiatedInternalClassInNamespace()
        {
            VerifyCSharp(
@"namespace N
{
    internal class C { }
}",
                GetCSharpResultAt(3, 20, AvoidUninstantiatedInternalClassesAnalyzer.Rule, "C"));
        }
```

Is this not what we want?

mavasani replied:
Yes, it is. Sorry, I didn't see the format properly and assumed it was emitting fully qualified names. Please ignore this comment.

mavasani commented on src/Microsoft.Maintainability.Analyzers/Core/AvoidUninstantiatedInternalClasses.cs:
```diff
@@ -37,6 +43,205 @@ public override void Initialize(AnalysisContext analysisContext)
 
             // TODO: Configure generated code analysis.
             //analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
```
You should uncomment this line and configure it with GeneratedCodeAnalysisFlags.Analyze - basically signup to receive callbacks for generated code (so we can note the type instantiations in generated code), but don't report diagnostics on types defined in generated code.

lgolding replied:
Done.

mavasani commented on src/Microsoft.Maintainability.Analyzers/Core/AvoidUninstantiatedInternalClasses.cs:
```diff
@@ -37,6 +43,205 @@ public override void Initialize(AnalysisContext analysisContext)
 
```
We should try to enable concurrent execution for this analyzer - basically uncomment the line `//analysisContext.EnableConcurrentExecution();` and then use concurrent sets instead of plain lists to hold instantiatedTypes and internalTypes.

lgolding replied:
Done. Thanks, TIL about `System.Collections.Concurrent`.

mavasani commented on src/Microsoft.Maintainability.Analyzers/Core/AvoidUninstantiatedInternalClasses.cs:
```diff
@@ -37,6 +43,205 @@ public override void Initialize(AnalysisContext analysisContext)
+            ISymbol assemblySymbol = compilation.Assembly;
+            ImmutableArray<AttributeData> attributes = assemblySymbol.GetAttributes();
+            return attributes.Any(
+                attr => attr.AttributeClass.Name.Equals("InternalsVisibleToAttribute", StringComparison.Ordinal));
```
We want to match against fully qualified name - basically define a well known attribute type like [this](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Analyzer.Utilities/WellKnownTypes.cs#L199), and just check if attr.AttributeClass equals that symbol.

lgolding replied:
Done.

mavasani commented on src/Microsoft.Maintainability.Analyzers/Core/AvoidUninstantiatedInternalClasses.cs:
```diff
@@ -37,6 +43,205 @@ public override void Initialize(AnalysisContext analysisContext)
+            }
+
+            // Attributes are not instantiated in IL but are created by reflection.
+            INamedTypeSymbol attributeSymbol = compilation.GetTypeByMetadataName("System.Attribute");
```
Again, we should get this well known symbol in the compilation start action, and pass it as an argument to this method.

lgolding replied:
Done.

mavasani commented on src/Microsoft.Maintainability.Analyzers/Core/AvoidUninstantiatedInternalClasses.cs:
```diff
@@ -37,6 +43,205 @@ public override void Initialize(AnalysisContext analysisContext)
+
+            // Types implementing the (deprecated) IConfigurationSectionHandler interface
+            // are OK because they are instantiated by the configuration system.
+            INamedTypeSymbol iConfigurationSectionHandlerSymbol = compilation.GetTypeByMetadataName("System.Configuration.IConfigurationSectionHandler");
```
Same here - in fact GetTypeByMetadataName should only be used in the compilation start action.

lgolding replied:
Done everywhere.

mavasani commented:
👍 modulo the suggestions to enable concurrent execution, configure generated code analysis and moving all GetTypeByMetadataName calls to compilation start action.

lgolding commented:
@mavasani Addressed all feedback, ready for final sign-off.

mavasani commented:
👍 Thanks Larry!

lgolding commented:
*phew*! Merged.
